lib/apiservers/engine/backends/eventmonitor.go

```diff
@@ -0,0 +1,217 @@
+	defer trace.End(trace.Begin(""))
+
+	plClient := PortLayerClient()
+	if plClient == nil {
+		return InternalServerError("eventproxy.StreamEvents failed to get a portlayer client")
+	}
+
+	params := events.NewGetEventsParamsWithContext(ctx)
+	if _, err := plClient.Events.GetEvents(params, out); err != nil {
+		switch err := err.(type) {
+		case *events.GetEventsInternalServerError:
+			return InternalServerError("Server error from the events port layer")
+		default:
+			//Check for EOF.  Since the connection, transport, and data handling are
+			//encapsulated inside of Swagger, we can only detect EOF by checking the
+			//error string
+			if strings.Contains(err.Error(), swaggerSubstringEOF) {
+				return nil
+			}
+			return InternalServerError(fmt.Sprintf("Unknown error from the interaction port layer: %s", err))
+		}
+	}
+
+	return nil	
+}
+
+func NewPortlayerEventMonitor(proxy eventproxy, publisher eventpublisher) *PortlayerEventMonitor {
+	return &PortlayerEventMonitor{proxy: proxy, publisher: publisher}
+}
+
+// Start() starts the portlayer event monitoring
+func (m *PortlayerEventMonitor) Start() error {
+	defer trace.End(trace.Begin(""))
+
+	if m.stop != nil {
+		return fmt.Errorf("Portlayer event monitor: Already started")
+	}
+
+	m.stop = make(chan struct{})
+	go m.monitor()
+
+	return nil
+}
+
+// Stop() stops the portlayer event monitoring
+func (m *PortlayerEventMonitor) Stop() {
+	defer trace.End(trace.Begin(""))
+
+	if m.stop != nil {
+		close(m.stop)
+	}
+}
+
+// monitor() establishes a streaming connection to the portlayer's event
+// endpoint, decodes the results, translate it to Docker events if needed,
+// and publishes the event to Docker event subscribers.
+func (m *PortlayerEventMonitor) monitor() error {
+	defer trace.End(trace.Begin(""))
+
+	var wg sync.WaitGroup
+	errors := make(chan error, 2)
+
+	reader, writer := io.Pipe()
+	ctx, cancel := context.WithCancel(context.TODO())
+	// Start streaming events
+	wg.Add(1)
+	go func() {
+		var err error
+
+		defer wg.Done()
+
+		if err = m.proxy.StreamEvents(ctx, writer); err != nil {
+			if ctx.Err() != context.Canceled {
+				log.Warnf("Event streaming from portlayer returned: %#v", err)
+			}
+		}
+		if ctx.Err() == context.Canceled {
+			log.Infof("Event streaming from portlayer was cancelled")
+			return
+		}
+		errors <- err
+
+		writer.Close()
+		reader.Close()
+	}()
+
+	// Start decoding event stream json
+	wg.Add(1)
+	go func() {
+		var err error
+		var event plevents.BaseEvent
+
+		defer wg.Done()
+
+		decoder := json.NewDecoder(reader)
+		for decoder.More() {
+			if err = decoder.Decode(&event); err == nil {
+				m.publisher.PublishEvent(event)
+			}
+		}
+		errors <- err
+
+		reader.Close()
+		writer.Close()
+	}()
+
+
+	// Create a channel signaling when the waitgroup finishes
+	done := make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(errors)
+		close(done)
+	}()
+
+	select{
+		case <- done:
+			for err := range errors {
```

Comment:
There really isn't a need to cancel here because we know that if we get here, both goroutines have already exited.